SongSnap_TextImageUnfrozen.ipynb

```diff
@@ -0,0 +1,4054 @@
+        "\n",
+        "        # Generate attention masks\n",
+        "        attention_masks = [[1 if token != 0 else 0 for token in seq] for seq in tokenized_sequences]\n",
+        "\n",
+        "        # Convert tokenized sequences and attention masks to tensors and move to device\n",
+        "        batch_X_lyric = torch.tensor(tokenized_sequences, dtype=torch.long).to(device)\n",
+        "        attention_masks = torch.tensor(attention_masks, dtype=torch.long).to(device)\n",
+        "\n",
+        "        batch_y_genre = y_genre_train[start:end]\n",
+        "        batch_y_popularity = y_popularity_train[start:end]\n",
+        "        batch_y_danceability = y_danceability_train[start:end]\n",
+        "        batch_y_energy = y_energy_train[start:end]\n",
+        "\n",
+        "\n",
+        "        # Zero the gradients\n",
+        "        optimizer_sound.zero_grad()\n",
+        "        # optimizer_album.zero_grad()\n",
+        "        optimizer_spec.zero_grad()\n",
+        "\n",
+        "\n",
+        "        optimizer_genre.zero_grad()\n",
+        "        optimizer_popularity.zero_grad()\n",
+        "        optimizer_danceability.zero_grad()\n",
+        "        optimizer_energy.zero_grad()\n",
+        "\n",
+        "        # TODO: Create embeddings for text and image\n",
+        "        # Retrieve the four outputs given our embeddings\n",
+        "        sound_embedding = sound_model(batch_X_sound)\n",
+        "        # album_embedding = album_model(batch_X_album)\n",
+        "        spec_embedding = spec_model(batch_X_spec)\n",
+        "        text_embedding = text_model(batch_X_lyric, attention_masks)\n",
+        "\n",
+        "\n",
+        "        genre_output, popularity_output, danceability_output, energy_output = overall_model(sound_embedding,\n",
+        "                                                                        spec_embedding,\n",
+        "                                                                        text_embedding)\n",
+        "\n",
+        "        genre_output = genre_output.to(device)\n",
+        "        batch_y_genre = batch_y_genre.to(device)\n",
+        "\n",
+        "        loss_genre = criterion_genre(genre_output, batch_y_genre.long())\n",
+        "        loss_popularity = criterion_popularity(popularity_output.view(-1), batch_y_popularity)\n",
+        "        loss_danceability = criterion_danceability(danceability_output.view(-1), batch_y_danceability)\n",
+        "        loss_energy = criterion_energy(energy_output.view(-1), batch_y_energy)\n",
+        "\n",
+        "        # Backpropagate through the genre, popularity, danceability, and energy layers using their respective losses\n",
+        "        # We retain graph because we are using the computation graph for multiple backprops\n",
```

Comment:
what does the computation graph refer to? is it the backpropped recursive values or is it an actual graph

Reply:
@preethiprakash1 would be a good person to answer this. maybe we can discuss during our meeting